crates/sui-indexer/src/store/mod.rs

```diff
@@ -200,6 +203,61 @@ pub mod diesel_macro {
         }};
     }
 
+    #[macro_export]
+    macro_rules! insert_or_ignore_into {
+        ($table:expr, $values:expr, $conn:expr) => {{
+            use diesel::RunQueryDsl;
+            #[cfg(feature = "postgres-feature")]
+            {
+                diesel::insert_into($table)
+                    .values($values)
+                    .on_conflict_do_nothing()
+                    .execute($conn)
+                    .map_err(IndexerError::from)
+                    .context("Failed to write tx_senders to PostgresDB")?;
+            }
+            #[cfg(feature = "mysql-feature")]
+            #[cfg(not(feature = "postgres-feature"))]
+            {
+                diesel::insert_or_ignore_into($table)
+                    .values($values)
+                    .execute($conn)
+                    .map_err(IndexerError::from)
+                    .context("Failed to write tx_senders to MySQLDB")?;
+            }
+        }};
+    }
+
+    #[macro_export]
+    macro_rules! on_conflict_do_update {
+        ($table:expr, $values:expr, $target:expr, $pg_columns:expr, $mysql_columns:expr, $conn:expr) => {{
+            use diesel::ExpressionMethods;
+            use diesel::RunQueryDsl;
+            #[cfg(feature = "postgres-feature")]
+            {
+                diesel::insert_into($table)
+                    .values($values)
+                    .on_conflict($target)
+                    .do_update()
+                    .set($pg_columns)
+                    .execute($conn)?;
+            }
+            #[cfg(feature = "mysql-feature")]
+            #[cfg(not(feature = "postgres-feature"))]
+            {
+                for excluded_row in $values.iter() {
+                    let columns = $mysql_columns;
+                    diesel::insert_into($table)
```

Comment:
here we execute the insertion row by row, which naively seems not efficient, can we batch insert?

Reply:
Good observation @gegaowp , I did not find an equivalent `on conflict do update` implementation in diesel mysql which gives you access to `excluded` table the way postgres diesel does and hence this row by row insert. You are right that this is less efficient, although I don't know if there is an easy workaround. I think potential options are:
1. Update diesel mysql to support excluded like semantics like because mysql does indeed support it
2. Use raw query to do `on conflict do update` for mysql
I think #2 is what we can go with once we realize the perf impact from it